// mangadex-api/dist/api/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountResolver = void 0;
const Agent_1 = require("../Agent");
const error_1 = require("../lib/error");
const base_1 = require("./base");
class AccountResolver extends base_1.ApiBase {
    static async activate(code) {
        const { data } = await Agent_1.Agent.call(`account/activate/${code}`);
        if (data.result === 'error') {
            throw new error_1.ApiResponseError(data.errors[0]);
        }
        return true;
    }
    /**
     * @param username 1-64 characters
     * @param password 8-1024 characters
     * @param email
     */
    static async createAccount(username, password, email) {
        const { data: user } = await Agent_1.Agent.call('account/create', {
            method: 'POST'
        }, {
            username,
            password,
            email
        });
        if (user.result === 'error') {
            return user;
        }
        return user;
    }
    static async resendActivation(email) {
        const { data } = await Agent_1.Agent.call('account/activate/resend', {
            method: 'POST'
        }, {
            email
        });
        return data;
    }
    static async recover(email) {
        const { data } = await Agent_1.Agent.call('account/recover', { method: 'POST' }, { email });
        return data;
    }
    static async completeRecover(code, newPassword) {
        const { data } = await Agent_1.Agent.call(`account/recover/${code}`, {
            method: 'POST'
        }, { newPassword });
        return data;
    }
}
exports.AccountResolver = AccountResolver;
